frontend/src/core/codemirror/language/embedded-python.ts

```diff
@@ -0,0 +1,163 @@
+const CLOSE_BRACE = 125; // }
+const DELIMITER_LENGTH = 1;
+
+const MARK = { mark: `${PYTHON}Mark`, resolve: PYTHON };
+
+/**
+ * Define a Python code block parser for Markdown.
+ *
+ * @param pythonParser CodeMirror {@link Parser} for Python code
+ * @returns Markdown extension
+ */
+export function parsePython(
+  pythonParser: Parser,
+  isActivated: () => boolean,
+): MarkdownConfig {
+  const defineNodes = [
+    { name: PYTHON, style: tags.emphasis },
+    { name: `${PYTHON}Mark`, style: tags.processingInstruction },
+  ];
+
+  return {
+    defineNodes,
+    parseInline: [
+      {
+        name: PYTHON,
+        parse(cx: InlineContext, next: number, pos: number): number {
+          if (!isActivated()) {
+            return -1;
+          }
+
+          if (next !== OPEN_BRACE && next !== CLOSE_BRACE) {
+            return -1;
+          }
+
+          return cx.addDelimiter(
+            MARK,
+            pos,
+            pos + DELIMITER_LENGTH,
+            next === OPEN_BRACE,
+            next === CLOSE_BRACE,
+          );
+        },
+      },
+    ],
+    wrap: parseMixed((node) => {
+      if (!isActivated()) {
+        return null;
+      }
+
+      if (node.type.name !== PYTHON) {
+        return null;
+      }
+
+      const from = node.from + DELIMITER_LENGTH;
+      const to = node.to - DELIMITER_LENGTH;
+
+      if (from >= to || from < 0 || to < 0) {
+        return null;
+      }
+
+      return {
+        parser: pythonParser,
+        overlay: [{ from, to }],
+        extensions: [python().support],
+      };
+    }),
+  };
+}
+
+const languageData = python().language.data;
+
+/**
+ * Embedded Python completions
+ *
+ * @param isActivated Function that returns whether the extension should be activated
+ */
+export function embeddedPythonCompletions(
+  isActivated: () => boolean,
+): Extension {
+  return [
+    languageData.of({
+      autocomplete: (context: CompletionContext): CompletionResult | null => {
+        if (!isActivated()) {
+          return null;
+        }
+
+        // Check if the cursor is at a position where a variable can be inserted
+        const wordMatch = context.matchBefore(/\w*$/);
+        if (!context.explicit && !wordMatch) {
+          return null;
+        }
+
+        const filter = wordMatch?.text ?? "";
+
+        const options = getVariableCompletions(
+          store.get(variablesAtom),
+          new Set(),
+        );
+
+        return {
+          from: context.pos - filter.length,
+          options,
+          validFor: /^\w*$/,
+        };
+      },
+    }),
+  ];
+}
+
+/**
+ * Cheap variable completion source without having to go through the
+ * backend.
+ */
+export const variableCompletionSource = (
+  context: CompletionContext,
+): CompletionResult | null => {
+  // Check if we're inside a {} block by looking for an opening brace
+  const beforeCursor = context.state.doc.sliceString(0, context.pos);
+  const lastOpenBrace = beforeCursor.lastIndexOf("{");
```

Comment:
For an `f` string, two consecutive braces inserts a literal `{`, which should not trigger completion, eg:

<img width="318" alt="image" src="https://github.com/user-attachments/assets/9ae47bda-a3cc-4ba3-baa9-36e884866c14" />